File: generated_code/GeminiPro/Java/044_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 *
 * A splay tree is a self-balancing binary search tree with the additional
 * property that recently accessed elements are quick to access again. It performs
 * basic operations such as insertion, deletion, and search. When an element is
 * accessed, it is moved to the root of the tree through a series of rotations
 * in an operation called "splaying".
 */
public class SplayTree {

    /**
     * Inner class representing a node in the splay tree.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer value for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the given node y.
     * This is a helper function for the splay operation.
     *
     * @param y The node to be rotated.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Performs a left rotation on the given node x.
     * This is a helper function for the splay operation.
     *
     * @param x The node to be rotated.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. It brings the node with the given key
     * (or the last accessed node if the key is not found) to the root of the tree.
     * This implementation uses the top-down splaying technique.
     *
     * @param root The root of the tree to splay.
     * @param key The key to splay around.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node root, int key) {
        if (root == null || root.key == key) {
            return root;
        }

        // Dummy node to simplify linking of left and right subtrees
        Node header = new Node(0);
        Node leftTreeMax = header;
        Node rightTreeMin = header;

        while (true) {
            if (key < root.key) {
                if (root.left == null) break;
                // Zig-Zig (left-left)
                if (key < root.left.key) {
                    root = rightRotate(root);
                }
                if (root.left == null) break;
                // Link to the right tree
                rightTreeMin.left = root;
                rightTreeMin = root;
                root = root.left;
            } else if (key > root.key) {
                if (root.right == null) break;
                // Zig-Zig (right-right)
                if (key > root.right.key) {
                    root = leftRotate(root);
                }
                if (root.right == null) break;
                // Link to the left tree
                leftTreeMax.right = root;
                leftTreeMax = root;
                root = root.right;
            } else { // key == root.key
                break;
            }
        }

        // Reassemble the trees
        leftTreeMax.right = root.left;
        rightTreeMin.left = root.right;
        root.left = header.right;
        root.right = header.left;

        return root;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree remains unchanged. After insertion,
     * the new node (or the existing node with the key) becomes the new root.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // Case 1: The tree is empty.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. The new root will be the closest node to the key.
        root = splay(root, key);

        // Case 2: The key is already in the tree.
        if (root.key == key) {
            return; // No duplicates allowed in the set.
        }

        // Case 3: The key is not in the tree. Insert a new node at the root.
        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Searches for a key in the tree. If the key is found, it is splayed
     * to the root. If not found, the last accessed node on the search path is
     * splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        // After splaying, if the key exists, it will be at the root.
        if (root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is found, it is deleted and the tree is rebalanced.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the key to the root.
        root = splay(root, key);

        // If the key is not at the root after splaying, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to its root.
            // Since `key` is larger than any element in the left subtree, splaying
            // for `key` will effectively bring the max element to the root.
            Node newRoot = splay(leftSubtree, key);

            // The new root of the left subtree now has no right child.
            // Attach the original right subtree to it.
            newRoot.right = rightSubtree;
            root = newRoot;
        }
    }
}

File: generated_code/GeminiPro/Python/044_SplayTree.py
class SplayTree:
    """
    A complete, self-contained Python class that implements a dictionary-like
    set for integers using a Splay Tree.

    The splay tree is a self-balancing binary search tree with the additional
    property that recently accessed elements are quick to access again.
    It performs basic operations such as insertion, look-up and removal
    in O(log n) amortized time.
    """

    class _Node:
        """A private inner class to represent a node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _rotate_right(self, y):
        """
        Performs a right rotation around the given node y.
        Returns the new root of the rotated subtree.
        """
        x = y.left
        y.left = x.right
        x.right = y
        return x

    def _rotate_left(self, x):
        """
        Performs a left rotation around the given node x.
        Returns the new root of the rotated subtree.
        """
        y = x.right
        x.right = y.left
        y.left = x
        return y

    def _splay(self, key):
        """
        Brings the node with the given key to the root of the tree.
        If the key is not in the tree, the last accessed node is brought
        to the root. This is the core top-down splay operation.
        """
        if not self.root:
            return

        # Create a dummy node to simplify linking of left and right subtrees
        dummy = self._Node(None)
        left_tree_max = dummy
        right_tree_min = dummy
        current = self.root

        while True:
            if key < current.key:
                if not current.left:
                    break
                # Zig-Zig case: perform a rotation if the key is also in the left-left grandchild
                if key < current.left.key:
                    current = self._rotate_right(current)
                    if not current.left:
                        break
                # Link current node to the right tree (nodes larger than key)
                right_tree_min.left = current
                right_tree_min = current
                current = current.left
            elif key > current.key:
                if not current.right:
                    break
                # Zig-Zig case: perform a rotation if the key is also in the right-right grandchild
                if key > current.right.key:
                    current = self._rotate_left(current)
                    if not current.right:
                        break
                # Link current node to the left tree (nodes smaller than key)
                left_tree_max.right = current
                left_tree_max = current
                current = current.right
            else:  # key == current.key
                break

        # Reassemble the tree
        left_tree_max.right = current.left
        right_tree_min.left = current.right
        current.left = dummy.right
        current.right = dummy.left

        self.root = current

    def insert(self, key):
        """
        Inserts a key into the tree. If the key already exists, the tree
        is splayed on that key but the set is not modified.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the tree on the key to bring the closest node to the root
        self._splay(key)

        # If key is already present, do nothing further
        if self.root.key == key:
            return

        # Create the new node and split the tree
        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            self.root.right = None

        self.root = new_node

    def search(self, key):
        """
        Searches for a key in the tree.
        This operation splays the accessed node (or its parent if not found)
        to the root.
        Returns True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        self._splay(key)

        return self.root.key == key

    def delete(self, key):
        """
        Deletes a key from the tree. If the key is not found, the tree
        is splayed on the last accessed node and no deletion occurs.
        """
        if not self.root:
            return

        self._splay(key)

        # If key is not in the tree after splaying, do nothing
        if self.root.key != key:
            return

        # The node to delete is now at the root
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left child, the right child becomes the new root
            self.root = right_subtree
        else:
            # Make the left subtree the new tree
            self.root = left_subtree
            # Splay on the original key again. This will bring the largest element
            # in the left subtree (the predecessor of the deleted key) to the root.
            self._splay(key)
            # Attach the original right subtree to the new root
            self.root.right = right_subtree